blivet/zfcp.py

```diff
@@ -47,62 +49,218 @@ def logged_write_line_to_file(fn, value):
 
     # Force str and unicode types in case any of the properties are unicode
     def _to_string(self):
-        return "%s %s %s" % (self.devnum, self.wwpn, self.fcplun)
+        return str(self.devnum)
 
     def __str__(self):
         return stringize(self._to_string())
 
     def __unicode__(self):
         return unicodeize(self._to_string())
 
-    def online_device(self):
-        online = "%s/%s/online" % (zfcpsysfs, self.devnum)
-        portadd = "%s/%s/port_add" % (zfcpsysfs, self.devnum)
-        portdir = "%s/%s/%s" % (zfcpsysfs, self.devnum, self.wwpn)
-        unitadd = "%s/unit_add" % (portdir)
-        unitdir = "%s/%s" % (portdir, self.fcplun)
-        failed = "%s/failed" % (unitdir)
+    def _free_device(self):
+        """Remove the device from the I/O ignore list to make it visible to the system.
 
-        if not os.path.exists(online):
+        :raises: ValueError if the device cannot be removed from the I/O ignore list
+        """
+
+        if not os.path.exists(self._device_online_path):
             log.info("Freeing zFCP device %s", self.devnum)
             util.run_program(["zfcp_cio_free", "-d", self.devnum])
 
-        if not os.path.exists(online):
+        if not os.path.exists(self._device_online_path):
             raise ValueError(_("zFCP device %s not found, not even in device ignore list.") %
                              (self.devnum,))
 
+    def _set_zfcp_device_online(self):
+        """Set the zFCP device online.
+
+        :raises: ValueError if the device cannot be set online
+        """
+
         try:
-            f = open(online, "r")
-            devonline = f.readline().strip()
-            f.close()
+            with open(self._device_online_path) as f:
+                devonline = f.readline().strip()
             if devonline != "1":
-                logged_write_line_to_file(online, "1")
+                logged_write_line_to_file(self._device_online_path, "1")
         except OSError as e:
             raise ValueError(_("Could not set zFCP device %(devnum)s "
                                "online (%(e)s).")
                              % {'devnum': self.devnum, 'e': e})
 
+    def _set_zfcp_device_offline(self):
+        """Set the zFCP device offline.
+
+        :raises: ValueError if the device cannot be set offline
+        """
+
+        try:
+            logged_write_line_to_file(self._device_online_path, "0")
+        except OSError as e:
+            raise ValueError(_("Could not set zFCP device %(devnum)s "
+                               "offline (%(e)s).")
+                             % {'devnum': self.devnum, 'e': e})
+
+    def _is_scsi_associated_with_fcp(self, fcphbasysfs, _fcpwwpnsysfs, _fcplunsysfs):
+        """Decide if the SCSI device with the provided SCSI attributes
+        corresponds to the zFCP device.
+
+        :returns: True or False
+        """
+
+        return fcphbasysfs == self.devnum
```

Comment:
Oh, hang on, shouldn't this method be defined in class `ZFCPDeviceAutoLunScan`, as opposed to this abstract `ZFCPDeviceBase`, which does not know whether auto lun scan is available or not?
But then again you seem to need it here to move method offline_scsi_device() to the base class so both derived classes can use it as common method.